sktime/forecasting/rbf_forecaster.py

```diff
@@ -0,0 +1,575 @@
+    r"""A lazy linear layer initialized based on the input dims.
+
+    This layer is useful when the input size is dynamic or unknown at the
+    initialization stage, enabling flexible architecture design.
+
+    Parameters
+    ----------
+    out_features : int
+        Number of output features for the linear layer.
+    """
+
+    def __init__(self, out_features):
+        super().__init__()
+        self.out_features = out_features
+        self.linear = None
+
+    def forward(self, x):
+        """Initialize and apply the linear layer if not already initialized.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor to be passed through the linear layer.
+
+        Returns
+        -------
+        torch.Tensor
+            Output tensor after linear transformation.
+        """
+        if self.linear is None:
+            self.linear = nn.Linear(x.shape[-1], self.out_features).to(x.device)
+        return self.linear(x)
+
+
+class RBFLayer(nn.Module):
+    r"""RBF layer to transform input data into a new feature space.
+
+    This layer applies an RBF transformation to each input feature,
+    expanding the feature space based on distances to predefined center points.
+
+    Parameters
+    ----------
+    in_features : int
+        Number of input features.
+    out_features : int
+        Number of output features (or RBF centers).
+    centers : array-like, optional (default=None)
+        The centers :math:`c_k` for the RBF transformation.
+        If None, centers are evenly spaced.
+    gamma : float, optional (default=1.0)
+        Parameter controlling the spread of the RBF.
+    rbf_type : {"gaussian", "multiquadric", "inverse_multiquadric"}
+                optional (default="gaussian")
+        The type of RBF kernel to apply.
+    """
+
+    def __init__(
+        self,
+        in_features,
+        out_features,
+        centers=None,
+        gamma=1.0,
+        rbf_type="gaussian",
+    ):
+        super().__init__()
+        self.in_features = in_features
+        self.out_features = out_features
+
+        if centers is None:
+            centers = torch.linspace(-1, 1, out_features).reshape(-1, 1)
+            centers = centers.repeat(1, in_features).numpy()
+
+        self.rbf_transformer = RBFTransformer(
+            centers=centers,
+            gamma=gamma,
+            rbf_type=rbf_type,
+            apply_to="values",
+            use_torch=True,
+        )
+
+    def forward(self, x):
+        """Apply the RBF transformation to the input data.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input data tensor.
+
+        Returns
+        -------
+        torch.Tensor
+            Transformed tensor with RBF features.
+        """
+        if len(x.shape) == 1:
+            x = x.unsqueeze(0)
+
+        x_np = x.detach().cpu().numpy()
+        self.rbf_transformer.fit(x_np)
+
+        x_rbf = self.rbf_transformer.transform(x_np)
```

Comment:
>How expensive is fitting an rbf_transformer? 

rbf_transformer is fitting is just creating centers. it creates its own centres if not provided using np.linspace.

>Thus, you might move it to the _fit method directly?

So, you want me to actually remove the classes `RBFLayer` etc? I added it as rn the forecaster has a predefined structure and the user may want only the layer to create his own "custom" forecaster... So, should I remove it and make the forecaster only?

Reply:
>stacking those layer is impossible...

I am just using one layer in `RBFNetwork` as I think that much is sufficient and then just adding linear layers. Are you saying that stacking one RBFLayer over another is impossible? So maybe yeah as once the data passes through it, it is transformed by the rbf_transformer and now the data is changed, I dont think we need to use this layer again on that data as it is already transformed and we got what we want.

Am I interpreting what you said correctly?

Reply:
So currently the transformation is done on each batch. I.e. also during predict this transformation is fitted and then applied. I am not familiar enough with the algorithm to know if it is correct that fit and transform is done one each  batch. 
Furthermore,transforming everything before starting training would be more efficient, since you do this right now for each batch multiple times..

Reply:
Ok, I will then just transform it only once before training

Reply:
Hi @benHeid, I had some doubts so I read a little more about RBF NNs and got to know about this:

A radial basis function (RBF) network contains three layers such as an input layer, a hidden layer with a non-linear [RBF](https://www.sciencedirect.com/topics/engineering/radial-basis-function) activation function and a linear output layer. The second layer /hidden layer perform a nonlinear mapping from the input space into a higher dimensional space by using a [Gaussian](https://www.sciencedirect.com/topics/physics-and-astronomy/gaussian-distribution) or some other kernel function. Output layer:-The final layer performs a weighted sum with a linear output.

(source: https://www.sciencedirect.com/topics/engineering/radial-basis-function-network)

I checked in multiple sites, everywhere they show the same structure. This is very similar to mine that I have already implemented. So, what should we do? Do the transformation only once or follow this structure?

The transformation does the same thing explained here: apply the RBF functions on the data

Reply:
lmk what do you think?